Fixed behaviour of S.Complexes in unions (#11730)

MrOverlord opened the pull request:
This PR fixes #11730

MrOverlord committed:
Fix union with complexes

MrOverlord committed:
Fix intersection with S.Reals

    Previous version failed on Intersection(ComplexRegion(Interval(0,1)**2), S.Reals)
    (expected [0,1], got S.Reals)

MrOverlord committed:
Fix test_union_RealSubSet not passing

MrOverlord committed:
Add unit test for issue 11730

    Modified test test_ComplexRegion_union to be consistent with other
    tests.

MrOverlord commented:
Oddly, the same travis checks have passed on my fork [link](https://travis-ci.org/MrOverlord/sympy/jobs/173728384).

jksuom commented:
That appears to be a Travis problem. I've restarted the job.

MrOverlord commented:
Any objections to merging?

jksuom commented on sympy/sets/fancysets.py:
```diff
@@ -1443,8 +1441,13 @@ def _union(self, other):
-        if self == S.Complexes:
-            return self
+        # treat a subset of Reals as complex numbers
+        if other.is_subset(S.Reals):
```
The method could probably check this first and coerce other into a ComplexRegion before asking `if other.is_ComplexRegion`.

MrOverlord replied:
I don't think converting other to ComplexRegion is necessary, since under the `if other.is_ComplexRegion:` we are only using `other.sets` which just returns set passed in constructor, and this way we don't incur overhead of ComplexRegion's constructor.

If you think this makes logic of what's happening clearer, I'll change it.

jksuom replied:
>  this makes logic of what's happening clearer

Yes, that is what I think. We should avoid having to use any knowledge of the internal representation of ComplexRegion. Hence also the conversion should be carried out by ComplexRegion despite of some overhead. That could be done by overloading `ComplexRegion.__new__` to accept a subset of Reals, or, maybe preferably, by a new method `ComplexRegion.from_real`.

MrOverlord committed:
Modify real subset handling

jksuom commented on sympy/sets/fancysets.py:
```diff
@@ -1433,6 +1431,11 @@ def _intersect(self, other):
+        if other.is_subset(S.Reals):
+
+            # treat a subset of reals as a complex region
+            other = ComplexRegion(other * FiniteSet(0), polar=self.polar)
```
`other*FiniteSet(0)` is a representation in rectangular coordinates. Hence polar should not be True.

MrOverlord replied:
Sympy allows negative radius in polar representation, so I thought allowing polar to be true would allow for more simplification (I'll change it to be consistent with mathematical definition).
```python
In [1]: rad, arg = Interval(-5,5), Interval(0, pi)

In [2]: C1 = ComplexRegion(rad * arg, polar=True)

In [3]: C1
Out[3]: {r⋅(ⅈ⋅sin(θ) + cos(θ)) | r, θ ∊ [-5, 5] × [0, π]}
```

MrOverlord committed:
Modify no longer allow real subset in polar form

    Consistent with mathematical definition.

MrOverlord commented:
@jksuom could you review please?

jksuom commented on sympy/sets/fancysets.py:
```diff
@@ -1433,6 +1431,11 @@ def _intersect(self, other):
     def _union(self, other):
 
+        if other.is_subset(S.Reals):
+
```
PEP-8 says: Use blank lines in functions, sparingly, to indicate logical sections.

jksuom commented:
There is a minor comment on style. Other than that this looks good to me.

MrOverlord committed:
Fix style

jksuom commented on sympy/sets/tests/test_fancysets.py:
```diff
@@ -593,8 +593,8 @@ def test_ComplexRegion_union():
-    assert c1.union(Interval(2, 4)) == Union(c1, Interval(2, 4), evaluate=False)
-    assert c5.union(Interval(2, 4)) == Union(c5, Interval(2, 4), evaluate=False)
+    assert c1.union(Interval(2, 4)) == Union.reduce(Union(c1, Interval(2, 4), evaluate=False).args)
+    assert c5.union(Interval(2, 4)) == Union.reduce(Union(c5, Interval(2, 4), evaluate=False).args)
```
These tests could rather have something like `Union(c1, ComplexRegion.from_real(Interval(2, 4)))` on the right hand side.

MrOverlord committed:
Add from_real method to ComplexRegion

    Also adapting tests to reflect behaviour:
    currently subsets of real numbers are converted to complex
    regions only in union with complex regions in rectangular
    coordinates (hence evaluate=False when testing union with
    set in polar coordinates).

jksuom commented:
This should be synchronized with the updated master.

MrOverlord committed:
Merge remote-tracking branch 'upstream/master' into issue11730

jksuom commented:
This 'upstream/master' does not seem to include the updated .travis.yml. git remote -v  should show 	git://github.com/sympy/sympy.git (fetch).

MrOverlord commented:
My remote -v contains the line `upstream	https://github.com/sympy/sympy.git (fetch)`
git log shows the following commit
```
commit a221c376f3f382d251a600aa336b45e95f92b7fe
Merge: 5827caf 4480598
Author: Kalevi Suominen <jksuom@gmail.com>
Date:   Tue Dec 13 09:40:05 2016 +0200

    Merge pull request #11935 from asmeurer/docs-fix
    
    Fix Sphinx docs build

```
Isn't that the commit with updated .travis.yml?

jksuom commented:
I was looking at the last commit above, which does not indicate any new updates. But apparently they never show on the github page.This is ok as commit a221c376 was included. The checks should now pass.

jksuom commented:
Thanks! I think this is ready to be merged.
BTW. This is not concerned with intersections other than in a test, but it seems that there might be a bug:
```
In [5]: ival = Interval(1, 2)

In [6]: sq = ComplexRegion(ival*unit)

In [7]: Intersection(S.Reals, sq)
Out[7]: ∅
```

MrOverlord commented:
It appears the issue is a simple typo in `intersect` method of ComplexRegion.
There's a similar bug for polar coordinates as well
```python
In [3]: rad, arg = Interval(0,1), FiniteSet(pi)

In [4]: cr = ComplexRegion(rad * arg, polar=True)

In [5]: Intersection(cr, S.Reals)
Out[5]: [0, 1]

```
Expected answer would be [-1,0]
